pkg/backend/snapshot.go

```diff
@@ -63,36 +65,60 @@ type SnapshotPersister interface {
 type SnapshotManager struct {
-	persister      SnapshotPersister    // The persister responsible for invalidating and persisting the snapshot
-	baseSnapshot   *deploy.Snapshot     // The base snapshot for this plan
-	secretsManager secrets.Manager      // The default secrets manager to use
-	resources      []*resource.State    // The list of resources operated upon by this plan
-	operations     []resource.Operation // The set of operations known to be outstanding in this plan
+	persister      SnapshotPersister // The persister responsible for invalidating and persisting the snapshot
+	baseSnapshot   *deploy.Snapshot  // The base snapshot for this plan
+	secretsManager secrets.Manager   // The default secrets manager to use
+	resources      []*resource.State // The list of resources operated upon by this plan
 
 	// The set of resources that have been operated upon already by this plan. These resources could also have
 	// been added to `resources` by other operations but need to be filtered out before writing the snapshot.
 	dones map[*resource.State]bool
 
-	completeOps      map[*resource.State]bool // The set of resources that have completed their operation
-	mutationRequests chan<- mutationRequest   // The queue of mutation requests, to be retired serially by the manager
-	cancel           chan bool                // A channel used to request cancellation of any new mutation requests.
-	done             <-chan error             // A channel that sends a single result when the manager has shut down.
+	completeOps map[*resource.State]bool // The set of resources that have completed their operation
+	cancel      chan bool                // A channel used to request cancellation of any new mutation requests.
+	done        <-chan error             // A channel that sends a single result when the manager has shut down.
 
 	refreshDeletes map[resource.URN]bool // The set of resources that have been deleted by a refresh in this plan.
-}
 
-var _ engine.SnapshotManager = (*SnapshotManager)(nil)
+	journalEvents chan JournalEntry
+	// The journal entries for this plan, which are used to replay the journal and produce a snapshot.
+	journalEntries JournalEntries
 
-type mutationRequest struct {
-	mutator func() bool
-	result  chan<- error
+	newResources gsync.Map[*resource.State, string]
 }
 
+var _ engine.SnapshotManager = (*SnapshotManager)(nil)
+
 func (sm *SnapshotManager) Close() error {
 	close(sm.cancel)
 	return <-sm.done
 }
 
+type JournalEntryKind int
+
+const (
+	JournalEntryBegin   JournalEntryKind = 0
+	JournalEntrySuccess JournalEntryKind = 1
+	JournalEntryFailure JournalEntryKind = 2
+	JournalEntryOutputs JournalEntryKind = 4
+)
+
+type JournalEntry struct {
+	Kind          JournalEntryKind
+	OperationUUID string // The UUID of the operation that this journal entry is associated with.
+	DeleteOld     int
+	DeleteNew     string              // UUID for the delete Operation
```

Comment:
As commented by Pat, we should just used indices not uuids here, will save bandwidth and the server can generate unique indices just as easy as uuids (just an atomic increment counter)